### PR TITLE
Update types.md

### PR DESCRIPTION
**Pull request by Reasoning-Technology**
The language was used that type is instantiated to create an instance.   It was explained that Julia instances are different than, say, objects in C++.  In some places the language of instances was used.  In some cases the language of objects.  I brought this up on the discussion forum, and suggested seeing what this would look like if the language of instances was used consistently.  Someone mentioned it sounded reasonable.  I am curious to see it, so here is a version uses the instance language consistently.

### CODE REVIEW

**Comment by JeffBezanson** [comment]
I don't think `object` and `instance` are direct synonyms. When talking about "a thing in a program" it's common to call it an "object", but when talking about it in the context of its type you're more likely to say "an instance of type T". Also the phrase "immutable instance" sounds very odd to me (unless perhaps followed by "of type T"); "immutable object" seems like a much better and more common term. So I definitely don't think the word "object" should be replaced by "instance" in every...instance.

A more important point of consistency might be "object" versus "value". It would be defensible to replace "value" with "object" in almost all cases. I think we avoided that because "value" seems a bit more general --- it might not be clear that even e.g. `2` is an "object" in julia.

**Comment by Antigonus** [comment]
JeffBazanson,  a 'value' is the result of applying a 'representation'  to an instance.  A value is not an 'object'.  Representations typically have a human aspect to them.   There are many standard representations, ANSI, IEEE, etc. for finding the values of different types of instances.  The concept of value predates that of object oriented programming by a long long time.

**Comment by Antigonus** [comment]
I'm not saying that an object and an instance are direct synonyms.

The term object is a semantic entity belong to the ontology of object oriented programming.

An 'instance' is contiguous memory that is bound to type.  Most programmers are accustom to instances allocated off heaps, as typed allocation off the heap is a program often available to programmers.   However, instances also occur on the stack and in data areas, or even on disk and in databases.  One might even say, on paper when you write a word and expect someone to read it back in English.   Instances are much older than objects of course.

**Comment by JeffBezanson** [comment]
> The concept of value predates that of object oriented programming by a long long time.

Right, I agree, which is why I said it would be ok to replace "value" with "object" in most cases --- every "thing" in a julia program is an object. A statement like "this object represents the value 2" makes sense.

**Comment by Antigonus** [comment]
JeffBezanson,  ah no,  value is the result of applying a representation to an instance.  A representation is a function in the mathematical sense of function.   This is very different from data that accepts messages - i.e. objects.   The concept of value is fundamental.  The conflation of the two concepts causes the documentation in this section to be a bit confusing.

**Comment by Antigonus** [comment]
The fundamental thing about type is that it can be instantiated.   It is not 'objectified'.   

In C++ an object is an instance bound to functions.   This  is close to Knuth's concept of abstract datatype.   In Julia, as for the Common Lisp Object System, you have not bound the instance to functions, rather you have introduced a multiple dispatch model, and kept instances separate.  Hence the correct language in many places where you have written 'object',  is actually 'instance'.  You don't intend any particular function binding, you intend to instantiate a type.   You instantiated a type, you get an instance.  You don't get an object until you bind functions to instances.

**Comment by JeffBezanson** [comment]
I don't know why you still think we disagree about something there. Given a thing in a program, knowing its representation tells you what value it...represents.

> This is very different from data that accepts messages - i.e. objects.

"Accepting messages" is only one interpretation of object-oriented programming (which itself does not really have a crisp definition). Notice that it's called the Common Lisp _Object_ System, not the Common Lisp Instance System. Anyway I think you're being uselessly pedantic --- these nuances are not terribly helpful to somebody reading the manual.

File: doc/src/manual/types.md [hunk 1]
```diff
@@ -1,15 +1,39 @@
...
+converts the contents of an instance into a *value*.
+
+Computer language type systems have traditionally been said to be either *static*, or
+*dynamic*.  In static type systems the types for all future memory instances are
+determined, and checks for their consistent use are made, before the program executable is
+run. This determination is made by the compiler that produces the executable.  The
+compiler will issue errors and refuse to generate code if it can not make the required
+determinations.  In contrast, in dynamic type systems the types are computed at run time
+typically just before computing the value to be held in the instance. This computation is
+performed by an interpreter.  Should the interpreter fail to determine a type or discover
```
[the remark this answers is not in the document]
Comment (StefanKarpinski) [reply doc/src/manual/types.md]: The word "traditionally" is not really the issue. The old text never states or implies that dynamic languages must be interpreted, whereas your version does, among a number of other things that are not accurate.

**Pull request closed without merging**